File: solutions/2111. Minimum Operations to Make the Array K-Increasing/2111.cpp
class Solution {
 public:
  int kIncreasing(vector<int>& arr, int k) {
    int ans = 0;

    for (int i = 0; i < k; ++i) {
      vector<int> arr;
      for (int j = i; j < arr.size(); j += k)
        arr.push_back(arr[j]);
      ans += numReplaced(arr);
    }

    return ans;
  }

 private:
  int numReplaced(const vector<int>& arr) {
    vector<int> tails;
    for (const int a : arr)
      if (tails.empty() || tails.back() <= a)
        tails.push_back(a);
      else
        tails[firstGreater(tails, a)] = a;
    return arr.size() - tails.size();
  }

  int firstGreater(const vector<int>& arr, int target) {
    return ranges::upper_bound(arr, target) - arr.begin();
  }
};

File: solutions/2111. Minimum Operations to Make the Array K-Increasing/2111.py
class Solution:
  def kIncreasing(self, arr: list[int], k: int) -> int:
    def numReplaced(arr: list[int]) -> int:
      tails = []
      for a in arr:
        if not tails or tails[-1] <= a:
          tails.append(a)
        else:
          tails[bisect_right(tails, a)] = a
      return len(arr) - len(tails)

    return sum(numReplaced(arr[i::k]) for i in range(k))
